src/core/ext/xds/file_watcher_certificate_provider_factory.cc

```diff
@@ -71,66 +71,54 @@ std::string FileWatcherCertificateProviderFactory::Config::ToString() const {
   return absl::StrJoin(parts, "");
 }
 
-RefCountedPtr<FileWatcherCertificateProviderFactory::Config>
-FileWatcherCertificateProviderFactory::Config::Parse(const Json& config_json,
-                                                     grpc_error_handle* error) {
-  auto config = MakeRefCounted<FileWatcherCertificateProviderFactory::Config>();
-  if (config_json.type() != Json::Type::kObject) {
-    *error = GRPC_ERROR_CREATE("error:config type should be OBJECT.");
-    return nullptr;
-  }
-  std::vector<grpc_error_handle> error_list;
-  ParseJsonObjectField(config_json.object(), "certificate_file",
-                       &config->identity_cert_file_, &error_list, false);
-  ParseJsonObjectField(config_json.object(), "private_key_file",
-                       &config->private_key_file_, &error_list, false);
-  if (config->identity_cert_file_.empty() !=
-      config->private_key_file_.empty()) {
-    error_list.push_back(GRPC_ERROR_CREATE(
+const JsonLoaderInterface*
+FileWatcherCertificateProviderFactory::Config::JsonLoader(const JsonArgs&) {
+  static const auto* loader =
+      JsonObjectLoader<Config>()
+          .OptionalField("certificate_file", &Config::identity_cert_file_)
+          .OptionalField("private_key_file", &Config::private_key_file_)
+          .OptionalField("ca_certificate_file", &Config::root_cert_file_)
+          .OptionalField("refresh_interval", &Config::refresh_interval_)
+          .Finish();
+  return loader;
+}
+
+void FileWatcherCertificateProviderFactory::Config::JsonPostLoad(
+    const Json& json, const JsonArgs& /*args*/, ValidationErrors* errors) {
```

Comment:
side non-blocking question on the JSON API - I wonder why we look at the JSON instead of the parsed object in `PostLoad`

Reply:
Because the field in the parsed object will be an empty string in both of the following cases:
1. The field was not present or null in the JSON.
2. The field was present in the JSON but had an invalid value.

The logic here wants to add errors in case 1.  In case 2, we will already have added an error during the normal validation for the field, so we don't want to add a second error here.